### ISSUE TITLE
"You say Schemata, I say Schemas"

### ISSUE THREAD

**Issue by pchampin** [w3c/tpac2023-breakouts#8]
### Session description

A large variety of schema languages exist, defined inside or outside of W3C; to name a few: [RDF-Schema], [OWL], [SHACL], [ShEx], [XML-Schema], [JSON-Schema]... Each of these languages have been favored by different categories of users, who in turn ignore, neglect, sometimes even despise the other languages, deemed "too complicated", "less powerful" or simply "not fit for purpose".

It might be tempting to consider that any schema language is worth any other, and that the "best" one is a matter of technological preferences. We argue on the contrary that these languages differ in their core purpose, and should be seen as complementary rather than competitors. More precisely:
* ontology languages such as [RDF-Schema] and [OWL] focus on the **conceptual modelling** of the domain,
* shape languages such as [SHACL] and [ShEx] focus on the **logical modelling** of the data,
* structural schema languages such as [XML-Schema] and [JSON-Schema] focus of the **physical modelling** of exchange formats.

Sticking to one schema language to cover all these aspects is therefore suboptimal. Creating bridges between their user communities, to allow cross-fertilization and combined use, is a promising approach.

But it is also challenging, because it creates the need to maintain consistency across schemas at different levels. We will present different tools and methods that have been proposed to deal with this problem, and discuss the standardization opportunities in this area.

[RDF-Schema]: https://www.w3.org/TR/rdf-schema/
[OWL]: https://www.w3.org/TR/owl2-primer/
[SHACL]: https://www.w3.org/TR/shacl/
[ShEx]: http://shex.io/shex-primer/index.html
[XML-Schema]: https://www.w3.org/TR/xmlschema-1/
[JSON-Schema]: https://json-schema.org/specification.html

### Session goal

We will discuss the complementarity of various schema languages, and which tools are available (or missing...) to make them work together.

### Who can attend

Anyone may attend (Default)

### Session duration

60 minutes (Default)

### Estimated number of in-person attendees

Don't know (Default)

### Agenda, minutes, slides, etc.

- Agenda: @@
- Slides: @@
- Minutes: @@

**Comment by egekorkan** [comment]
As a follow-up, here are the takeaways:

- Minutes are available in the calendar link but [here](https://www.w3.org/2023/09/13-schemata-minutes.html) they are.
- As can be seen in the last part of the minutes, the discussion will follow up in a new community group since there is strong interest.
- My presentation is available at https://github.com/w3c/wot/blob/main/PRESENTATIONS/2023-09-tpac/2023-09-13-Breakout-Schemata-TD_Single_Source_of_Truth.pdf . I am ok with copying this presentation to a W3C location for posteriority.

@pchampin can you get the presentations from the other presenters? For the record, there were 4 presentations in total: Pierre-Antoine, me, TreeLDR and LinkML

**Comment by mristin** [comment]
Hi!
Please apologize for joining this discussion out of the blue. I do research in the field of Industry 4.0 and had recently a very interesting chat with @wiresio. It turned out that some of our work on I4.0 meta-models, generating schemas and SDKs could be transferred to your domain as well.

We formalize the meta-model in a subset of Python language, parse that meta-model into an intermediate representation, and, based on that, generate all the downstream artifacts (SDKs in different languages, JSON Schema, XSD, RDF+SHACL, *etc.*).

Probably it would be best if we schedule a phone conference and I present you the approach, but here are some entry points if you already want to have a look at what we did so far.

https://www.researchgate.net/publication/356039469_Generative_and_Model-driven_SDK_development_for_the_Industrie_40_Digital_Twin?_tp=eyJjb250ZXh0Ijp7ImZpcnN0UGFnZSI6Il9kaXJlY3QiLCJwYWdlIjoic2VhcmNoIiwicG9zaXRpb24iOiJwYWdlSGVhZGVyIn19

https://www.researchgate.net/publication/373325991_Generation_of_Digital_Twins_for_Information_Exchange_Between_Partners_in_the_Industrie_40_Value_Chain

https://www.researchgate.net/publication/374055610_Maturity_Evaluation_of_SDKs_for_I40_Digital_Twins

For more background on code generation, please contact me on my [email](mailto:rist@zhaw.ch), as the related paper has been submitted and accepted at IECON 2023, but not published yet.

A bit outdated repository of design docs is available here:
https://github.com/aas-core-works/design-docs

For a list of all the things we generate, see:
https://github.com/aas-core-works

The meta-models live in:
https://github.com/aas-core-works/aas-core-meta

The main generator lives in:
https://github.com/aas-core-works/aas-core-codegen

**Comment by wiresio** [comment]
Thanks @mristin for joining the discussion!

As you wrote, it would be absolutly interesting to learn from your experiences!

@pchampin, @egekorkan - what do you think?

**Comment by mristin** [comment]
> Thanks @mristin for the inputs. Your use case is also relevant. Our decision in TPAC was to propose a new community group such that these discussions can be had together. Would you be interested in participating?

Certainly! Best you contact me via email or via Linkedin for the next steps so that we do not spam this issue.

**Comment by VladimirAlexiev** [comment]
@cmungall @pchampin @egekorkan 
I'm quite interested in such CG.
See https://github.com/json-ld/yaml-ld/issues/19 where I've collected a list of frameworks/cases for what I call "Polyglot Modeling", and pinged a bunch of people with potential interest.

I've subscribed for https://github.com/w3c/breakouts-day-2024/issues/15

@mristin and @wiresio I'm very interested in AAS use cases based on semantic technologies.
And very interested in applications of Electrical CIM: I think that's also a good area for code generation based on models.
Maybe contact me `vladimir dot alexiev at ontotext dot com` or https://www.linkedin.com/in/valexiev for discussion?

**Comment by wiresio** [comment]
@VladimirAlexiev - Always happy to brainstorm about new ideas! Will send you an email for this. Would also be great if you could join the breakout session mentioned above.
